Flow.Launcher/Storage/TopMostRecord.cs

```diff
@@ -1,29 +1,29 @@
-﻿using System.Collections.Generic;
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.Text.Json.Serialization;
 using Flow.Launcher.Plugin;
 
 namespace Flow.Launcher.Storage
 {
-    // todo this class is not thread safe.... but used from multiple threads.
     public class TopMostRecord
     {
         [JsonInclude]
-        public Dictionary<string, Record> records { get; private set; } = new Dictionary<string, Record>();
+        public ConcurrentDictionary<string, Record> records { get; private set; } = new ConcurrentDictionary<string, Record>();
 
         internal bool IsTopMost(Result result)
         {
-            if (records.Count == 0 || !records.ContainsKey(result.OriginQuery.RawQuery))
+            if (records.IsEmpty || (result.OriginQuery != null && !records.ContainsKey(result.OriginQuery.RawQuery)))
```

Comment:
consider separate the contains key to later logic so you don't have to repeat checking the `result.OriginQuery`

Reply:
Get it!